features/dd-sdk-android-rum/src/test/kotlin/com/datadog/android/rum/internal/metric/RumSessionEndedMetricTest.kt

```diff
@@ -0,0 +1,412 @@
+            "top3" to 8,
+            "top4" to 7,
+            "top5" to 6
+        )
+        assertEquals(expectedErrorKind, sdkErrorCount[SessionEndedMetric.SDK_ERRORS_COUNT_BY_KIND_KEY])
+    }
+
+    @Test
+    fun `M error kind escape non alpha numeric W track errors`() {
+        // Given
+        val errorKindsMap = mapOf(
+            "top 1 error" to 10,
+            "top:2:error" to 9,
+            "top_3_error" to 8,
+            "top/4/error" to 7
+        )
+        val sessionEndedMetric = SessionEndedMetric(fakeSessionId, fakeStartReason)
+
+        // When
+        errorKindsMap.forEach { errorKind ->
+            repeat(errorKind.value) {
+                sessionEndedMetric.onErrorTracked(errorKind.key)
+            }
+        }
+
+        // Then
+        val attributes = sessionEndedMetric.toMetricAttributes()
+        val rse = attributes[SessionEndedMetric.RSE_KEY] as Map<*, *>
+        val sdkErrorCount = rse[SessionEndedMetric.SDK_ERRORS_COUNT_KEY] as Map<*, *>
+
+        val expectedErrorKind = mapOf(
+            "top_1_error" to 10,
+            "top_2_error" to 9,
+            "top_3_error" to 8,
+            "top_4_error" to 7
+        )
+        assertEquals(expectedErrorKind, sdkErrorCount[SessionEndedMetric.SDK_ERRORS_COUNT_BY_KIND_KEY])
+    }
+
+    @Test
+    fun `M encode type W creating metric`(@Forgery viewEvent: ViewEvent) {
+        // Given
+        val sessionEndedMetric = SessionEndedMetric(fakeSessionId, fakeStartReason)
+
+        // When
+        viewEvent.copy(session = viewEvent.session.copy(id = fakeSessionId)).apply {
+            sessionEndedMetric.onViewTracked(this)
+        }
+
+        // Then
+
+        val attributes = sessionEndedMetric.toMetricAttributes()
+        val json = JSONObject(attributes)
+
+        val rseObject = json.get(SessionEndedMetric.RSE_KEY) as JSONObject
+        Assertions.assertNotNull(rseObject.get(SessionEndedMetric.PROCESS_TYPE_KEY) as String)
+        Assertions.assertNotNull(rseObject.get(SessionEndedMetric.PRECONDITION_KEY) as String)
+        Assertions.assertNotNull(rseObject.get(SessionEndedMetric.DURATION_KEY) as Long)
+        Assertions.assertNotNull(rseObject.get(SessionEndedMetric.WAS_STOPPED_KEY) as Boolean)
+
+        val viewCountsObject = rseObject.get(SessionEndedMetric.VIEW_COUNTS_KEY) as JSONObject
+        Assertions.assertNotNull(viewCountsObject.get(SessionEndedMetric.VIEW_COUNTS_TOTAL_KEY) as Int)
+        Assertions.assertNotNull(viewCountsObject.get(SessionEndedMetric.VIEW_COUNTS_BG_KEY) as Int)
+        Assertions.assertNotNull(viewCountsObject.get(SessionEndedMetric.VIEW_COUNTS_APP_LAUNCH_KEY) as Int)
+
+        val sdkErrorsCount = rseObject.get(SessionEndedMetric.SDK_ERRORS_COUNT_KEY) as JSONObject
+        Assertions.assertNotNull(sdkErrorsCount.get(SessionEndedMetric.SDK_ERRORS_COUNT_TOTAL_KEY) as Int)
+        Assertions.assertNotNull(sdkErrorsCount.get(SessionEndedMetric.SDK_ERRORS_COUNT_BY_KIND_KEY) as JSONObject)
```

Comment:
it cannot be null, because you already cast to the non-nullable type. If value is null, it would fail earlier, with `ClassCastException`, so I guess you need to modify these lines.

Reply:
Do you think change `as` to `as?` will do the trick? If the value is not mapped to the desired type, it will become null, then we can check its nullability.

Reply:
yes, it will. but you can also use `assertThat(...).isInstanceOf(...)` (or maybe even more descriptive assertion exists), which will be even more descriptive.

Reply:
Good point, updated